### ISSUE TITLE
WSREP_SST_RSYNC not syncing with >10.2

### ISSUE THREAD

**Issue by jdel** [docker-library/mariadb#126]
Hello, 

I am experiencing issues to get a mariadb galera cluster going. For the sake of keeping it short, I will only start 2 hosts here.

I bootstrap an empty cluster on host1 (192.168.99.31) using the following:

```
#!/bin/bash
docker run -d \
--user mysql \
--name testgalera1 \
-h host1 \
--publish "3307:3307" \
--publish "4444:4444" \
--publish "4567:4567" \
--publish "4568:4568" \
--volume /data/mysql/:/var/lib/mysql/:rw \
-e TERM=xterm \
-e MYSQL_INITDB_SKIP_TZINFO=yes \
-e MYSQL_ROOT_PASSWORD=admin \
mariadb:10.2 \
--port=3307 \
--wsrep_debug="ON" \
--wsrep-new-cluster \
--wsrep-cluster-name="rancher" \
--wsrep-cluster-address=gcomm:// \
--wsrep-on="ON" \
--wsrep-provider="/usr/lib/galera/libgalera_smm.so" \
--wsrep-node-address="192.168.99.31" \
--wsrep-node-name="host1" \
--binlog-format=row \
--default-storage-engine=InnoDB \
--innodb-autoinc-lock-mode=2 \
--innodb-flush-log-at-trx-commit=0 \
--query-cache-size=0 \
--query-cache-type=0 \
--max-connections=300 \
&& docker logs -tf testgalera1
```

Everything's good, the last lines in the log read:

```
2017-08-30T09:28:50.087532189Z Version: '10.1.26-MariaDB-1~jessie'  socket: '/var/run/mysqld/mysqld.sock'  port: 3307  mariadb.org binary distribution
2017-08-30T09:28:50.096312184Z 2017-08-30  9:28:50 140274527681280 [Note] WSREP: Synchronized with group, ready for connections
2017-08-30T09:28:50.096340033Z 2017-08-30  9:28:50 140274527681280 [Note] WSREP: wsrep_notify_cmd is not defined, skipping notification.
2017-08-30T09:28:50.096345633Z 2017-08-30  9:28:50 140274527681280 [Note] WSREP: Nobody is waiting for SST.
```

I bring up the second container on host2 (192.168.99.32) with:

```
#!/bin/bash
docker run -d \
--user mysql \
--name testgalera2 \
-h host2 \
--publish "3307:3307" \
--publish "4444:4444" \
--publish "4567:4567" \
--publish "4568:4568" \
--volume /data/mysql/:/var/lib/mysql/:rw \
-e TERM=xterm \
mariadb:10.2 \
--port=3307 \
--wsrep_debug="ON" \
--wsrep-new-cluster \
--wsrep-cluster-name="rancher" \
--wsrep-cluster-address=gcomm:// \
--wsrep-on="ON" \
--wsrep-provider="/usr/lib/galera/libgalera_smm.so" \
--wsrep-node-address="192.168.99.32" \
--wsrep-node-name="host2" \
--binlog-format=row \
--default-storage-engine=InnoDB \
--innodb-autoinc-lock-mode=2 \
--innodb-flush-log-at-trx-commit=0 \
--query-cache-size=0 \
--query-cache-type=0 \
--max-connections=300 \
&& docker logs -tf testgalera2
```

The last lines of log from this container are:

```
2017-08-30T09:29:25.817020365Z 2017-08-30  9:29:25 140089686861568 [Warning] WSREP: Gap in state sequence. Need state transfer.
2017-08-30T09:29:25.817022929Z 2017-08-30  9:29:25 140089686861568 [Note] WSREP: Setting wsrep_ready to 0
2017-08-30T09:29:25.817025554Z 2017-08-30  9:29:25 140089350940416 [Note] WSREP: Running: 'wsrep_sst_rsync --role 'joiner' --address '192.168.99.32' --datadir '/var/lib/mysql/'   --parent '1'  '' '
2017-08-30T09:29:27.818079696Z 2017-08-30  9:29:27 140089574651648 [Note] WSREP: (b651ac4f, 'tcp://0.0.0.0:4567') turning message relay requesting off
```

Then nothing happens, the node never connects.

I spent a whole day trying every available wsrep option, different IPs, different VMs, different OS... to finally try the `mariadb:10.1` container... and my second node syncs perfectly.

Copying the 10.1 `wsrep_sst_rsync` script into the `mariadb:10.2` container works too, so I gathered debug output of the `wsrep_sst_rsync` scripts for both versions, to notice the following in the 10.2 version:

```
[TRUNCATED...]
+ is_local_ip 192.168.99.32
+ local address=192.168.99.32
+ which ifconfig
+ local get_addr_bin=
+ [ -z  ]
+ which ip
+ get_addr_bin=/sbin/ip
+ get_addr_bin=/sbin/ip address show
+ address=192.168.99.32/
+ grep 192.168.99.32/
+ /sbin/ip address show
+ RSYNC_ADDR=*
+ RSYNC_REAL_PID=2837
+ check_pid_and_port /var/lib/mysql//rsync_sst.pid 2837 bin boot dev docker-entrypoint-initdb.d docker-entrypoint.sh etc home lib lib64 media mnt opt proc root run sbin srv sys tmp usr var 4444
+ local pid_file=/var/lib/mysql//rsync_sst.pid
+ local rsync_pid=2837
+ local rsync_addr=bin
+ local rsync_port=boot
+ which lsof
+ rsync --daemon --no-detach --port 4444 --config /var/lib/mysql//rsync_sst.conf
+ grep (LISTEN)
+ lsof -i :boot -Pn
+ local port_info=
+ grep -w ^rsync[[:space:]]\+2837
+ echo
+ local is_rsync=
+ grep *:boot
+ echo
+ local is_listening_all=
+ grep bin:boot
+ echo
+ local is_listening_addr=
+ [ ! -z  -o ! -z  ]
+ check_pid /var/lib/mysql//rsync_sst.pid
+ local pid_file=/var/lib/mysql//rsync_sst.pid
+ [ -r /var/lib/mysql//rsync_sst.pid ]
+ cat /var/lib/mysql//rsync_sst.pid
+ ps -p 2837
+ [ -n  ]
+ sleep 0.2
+ check_pid_and_port /var/lib/mysql//rsync_sst.pid 2837 bin boot dev docker-entrypoint-initdb.d docker-entrypoint.sh etc home lib lib64 media mnt opt proc root run sbin srv sys tmp usr var 4444
+ local pid_file=/var/lib/mysql//rsync_sst.pid
+ local rsync_pid=2837
+ local rsync_addr=bin
+ local rsync_port=boot
+ which lsof
+ grep (LISTEN)
+ lsof -i :boot -Pn
+ local port_info=
+ grep -w ^rsync[[:space:]]\+2837
+ echo
+ local is_rsync=
+ grep *:boot
+ echo
+ local is_listening_all=
+ grep bin:boot
+ echo
+ local is_listening_addr=
+ [ ! -z  -o ! -z  ]
+ check_pid /var/lib/mysql//rsync_sst.pid
+ local pid_file=/var/lib/mysql//rsync_sst.pid
+ [ -r /var/lib/mysql//rsync_sst.pid ]
+ cat /var/lib/mysql//rsync_sst.pid
+ ps -p 2837
+ [ -n  ]
+ sleep 0.2
[TRUNCATED... it keeps on looping]
```

It loops on and on and on. But the interesting thing is that `check_pid_and_port` is called with way more parameters than it should (`bin boot dev docker-entrypoint-initdb.d docker-entrypoint.sh etc home lib lib64 media mnt opt proc root run sbin srv sys tmp usr var`).

This is because of `RSYNC_ADDR=*` just above that gets expanded when executing L360: `    until check_pid_and_port $RSYNC_PID $RSYNC_REAL_PID $RSYNC_ADDR $RSYNC_PORT`.

It is expected that that `RSYNC_ADDR` gets set to `*` in a docker container because `is_local_ip` will be false, but good practice recommends that variables should be double quoted. 

Double quoting the variables in the line above and restarting the container makes my second node complete SST successfully.

I am afraid I already wasted too much time troubleshooting this, but happy to make a PR to fix this particular line. In reality, this script should be reviewed entirely to quote every variable.

Until this gets fixed in 10.2+, I will be using 10.1.

# TL;DR

The `wsrep_sst_rsync` shipping with mariadb:10.2 container is not shell-safe enough and is missing a lot of double quotes around variables that can introduce unexpected behavior with `RSYNC_ADDR=*` for example.

**Comment by jdel** [comment]
PR follow up here https://github.com/MariaDB/server/pull/443

**Comment by tanji** [comment]
Interesting case, but not a docker issue, though. You did right reporting it upstream.
